### PR TITLE
Define show for CartesianIndices

### LINKED ISSUE #43750: code_warntype output is unmanageably verbose for CartesianIndex range
This function
```Julia
julia> function f(a)
           for i in CartesianIndex(1,1):CartesianIndex(10,10)
               a[i] = 0
           end
       end
```
leads to a very long output with `code_warntype`:
```Julia
julia> code_warntype(f, (Matrix{Float64},))
MethodInstance for f(::Matrix{Float64})
  from f(a) in Main at REPL[1]:1
Arguments
  #self#::Core.Const(f)
  a::Matrix{Float64}
Locals
  @_3::Union{Nothing, Tuple{CartesianIndex{2}, CartesianIndex{2}}}
  i::CartesianIndex{2}
Body::Nothing
1 ─ %1  = Main.CartesianIndex(1, 1)::Core.Const(CartesianIndex(1, 1))
│   %2  = Main.CartesianIndex(10, 10)::Core.Const(CartesianIndex(10, 10))
│   %3  = (%1:%2)::Core.Const(CartesianIndex{2}[CartesianIndex(1, 1) CartesianIndex(1, 2) CartesianIndex(1, 3) CartesianIndex(1, 4) CartesianIndex(1, 5) CartesianIndex(1, 6) CartesianIndex(1, 7) CartesianIndex(1, 8) CartesianIndex(1, 9) CartesianIndex(1, 10); CartesianIndex(2, 1) CartesianIndex(2, 2) CartesianIndex(2, 3) CartesianIndex(2, 4) CartesianIndex(2, 5) CartesianIndex(2, 6) CartesianIndex(2, 7) CartesianIndex(2, 8) CartesianIndex(2, 9) CartesianIndex(2, 10); CartesianIndex(3, 1) CartesianIndex(3, 2) CartesianIndex(3, 3) CartesianIndex(3, 4) CartesianIndex(3, 5) CartesianIndex(3, 6) CartesianIndex(3, 7) CartesianIndex(3, 8) CartesianIndex(3, 9) CartesianIndex(3, 10); CartesianIndex(4, 1) CartesianIndex(4, 2) CartesianIndex(4, 3) CartesianIndex(4, 4) CartesianIndex(4, 5) CartesianIndex(4, 6) CartesianIndex(4, 7) CartesianIndex(4, 8) CartesianIndex(4, 9) CartesianIndex(4, 10); CartesianIndex(5, 1) CartesianIndex(5, 2) CartesianIndex(5, 3) CartesianIndex(5, 4) CartesianIndex(5, 5) CartesianIndex(5, 6) CartesianIndex(5, 7) CartesianIndex(5, 8) CartesianIndex(5, 9) CartesianIndex(5, 10); CartesianIndex(6, 1) CartesianIndex(6, 2) CartesianIndex(6, 3) CartesianIndex(6, 4) CartesianIndex(6, 5) CartesianIndex(6, 6) CartesianIndex(6, 7) CartesianIndex(6, 8) CartesianIndex(6, 9) CartesianIndex(6, 10); CartesianIndex(7, 1) CartesianIndex(7, 2) CartesianIndex(7, 3) CartesianIndex(7, 4) CartesianIndex(7, 5) CartesianIndex(7, 6) CartesianIndex(7, 7) CartesianIndex(7, 8) CartesianIndex(7, 9) CartesianIndex(7, 10); CartesianIndex(8, 1) CartesianIndex(8, 2) CartesianIndex(8, 3) CartesianIndex(8, 4) CartesianIndex(8, 5) CartesianIndex(8, 6) CartesianIndex(8, 7) CartesianIndex(8, 8) CartesianIndex(8, 9) CartesianIndex(8, 10); CartesianIndex(9, 1) CartesianIndex(9, 2) CartesianIndex(9, 3) CartesianIndex(9, 4) CartesianIndex(9, 5) CartesianIndex(9, 6) CartesianIndex(9, 7) CartesianIndex(9, 8) CartesianIndex(9, 9) CartesianIndex(9, 10); CartesianIndex(10, 1) CartesianIndex(10, 2) CartesianIndex(10, 3) CartesianIndex(10, 4) CartesianIndex(10, 5) CartesianIndex(10, 6) CartesianIndex(10, 7) CartesianIndex(10, 8) CartesianIndex(10, 9) CartesianIndex(10, 10)])
│         (@_3 = Base.iterate(%3))
│   %5  = (@_3::Core.Const((CartesianIndex(1, 1), CartesianIndex(1, 1))) === nothing)::Core.Const(false)
│   %6  = Base.not_int(%5)::Core.Const(true)
└──       goto #4 if not %6
2 ┄ %8  = @_3::Tuple{CartesianIndex{2}, CartesianIndex{2}}
│         (i = Core.getfield(%8, 1))
│   %10 = Core.getfield(%8, 2)::CartesianIndex{2}
│         Base.setindex!(a, 0, i)
│         (@_3 = Base.iterate(%3, %10))
│   %13 = (@_3 === nothing)::Bool
│   %14 = Base.not_int(%13)::Bool
└──       goto #4 if not %14
3 ─       goto #2
4 ┄       return nothing
```

Note that every index combination is listed explicitly for `%3`. There doesn't seem to be a limit for this, and output for a 3d loop with 128^3 elements is unmanageable. `kill -9` is the only way out.

This is with both Julia 1.7 and the current master branch.

### PR DESCRIPTION
**Pull request by eschnett**
Closes https://github.com/JuliaLang/julia/issues/43750.

### CODE REVIEW

**Comment by simeonschaub** [comment]
Maybe even use the more terse printing `CartesianIndices((1:2:5, 1:1:2))`?

**Comment by eschnett** [comment]
Can do.

**Review by simeonschaub — APPROVED** [review]
Thanks! Looks great!

File: base/multidimensional.jl [hunk 1]
```diff
@@ -203,9 +203,7 @@ module IteratorsMD
     CartesianIndex(2, 2, 2)
 
     julia> CartesianIndices(fill(1, (2,3)))
-    2×3 CartesianIndices{2, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}}:
-     CartesianIndex(1, 1)  CartesianIndex(1, 2)  CartesianIndex(1, 3)
-     CartesianIndex(2, 1)  CartesianIndex(2, 2)  CartesianIndex(2, 3)
+    CartesianIndices((Base.OneTo(2), Base.OneTo(3)))
```
Comment (simeonschaub) [inline base/multidimensional.jl:206]: Not sure if it's worth the trouble, but we might want to even print that as `CartesianIndices((2, 3))` instead.

Comment (eschnett) [reply base/multidimensional.jl:206]: Unfortunately that wouldn't be accepted by the constructor. I like output formats that parse to produce the respective values again.

Comment (simeonschaub) [reply base/multidimensional.jl:206]: There is actually a corresponding constructor, that's what I meant:
```
julia> CartesianIndices((2, 3))
2×3 CartesianIndices{2, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}}:
 CartesianIndex(1, 1)  …  CartesianIndex(1, 3)
 CartesianIndex(2, 1)     CartesianIndex(2, 3)
```

Comment (eschnett) [reply base/multidimensional.jl:206]: What would the rule be? All `OneTo` values are printed as scalars instead?
```Julia
julia> CartesianIndices((2, 1:3))
CartesianIndices((Base.OneTo(2), 1:3))
```

Comment (simeonschaub) [reply base/multidimensional.jl:206]: I think the behavior for plain integers mixed with ranges changed at some point, so we might only want to do that if all values are `OneTo`.

Comment (eschnett) [reply base/multidimensional.jl:206]: Done.

**Comment by eschnett** [comment]
The build error is from the `Distributed` package and seems unrelated.

**Pull request merged**